aws/src/main/java/org/apache/iceberg/aws/s3/S3InputStream.java

```diff
@@ -88,23 +96,69 @@ public void seek(long newPos) {
 
   @Override
   public int read() throws IOException {
-    Preconditions.checkState(!closed, "Cannot read: already closed");
-    positionStream();
+    int[] byteRef = new int[1];
+    try {
+      Tasks.foreach(0)
+          .retry(awsProperties.s3ReadRetryNumRetries())
+          .exponentialBackoff(
+              awsProperties.s3ReadRetryMinWaitMs(),
+              awsProperties.s3ReadRetryMaxWaitMs(),
+              awsProperties.s3ReadRetryTotalTimeoutMs(),
+              2.0 /* exponential */)
+          .shouldRetryTest(S3InputStream::shouldRetry)
+          .throwFailureWhenFinished()
+          .run(ignored -> {
+            try {
+              Preconditions.checkState(!closed, "Cannot read: already closed");
+              positionStream();
+
+              byteRef[0] =  stream.read();
+            } catch (IOException e) {
+              closeStream();
+              throw new UncheckedIOException(e);
```

Comment:
on an unrecoverable network error, `abortStream()` is required to avoid the failed TLS connection being recycled. This adds a new challenge: identify an unrecoverable network error from the deep nested stacks which can come from the SDK, and distinguish from other failures